go/interpreter/lfvm/instructions.go

```diff
@@ -862,27 +862,26 @@ func opExtcodehash(c *context) {
 	}
 }
 
-func checkInitCodeSize(c *context, size *uint256.Int) bool {
+// checkInitCodeSize checks the size of the init code.
+// An error is returned if size is greater than MaxInitCodeSize, or
+// if not enough gas is available.
+// Caller should handle error return and stop the execution.
+func checkInitCodeSize(c *context, size *uint256.Int) error {
 	const (
 		MaxCodeSize     = 24576           // Maximum bytecode to permit for a contract
 		MaxInitCodeSize = 2 * MaxCodeSize // Maximum initcode to permit in a creation transaction and create instructions
 		InitCodeWordGas = 2               // Once per word of the init code when creating a contract.
 	)
 
 	if !c.isAtLeast(tosca.R12_Shanghai) {
-		return true
+		return nil
 	}
 	if !size.IsUint64() || size.Uint64() > MaxInitCodeSize {
 		c.useGas(c.gas)
 		c.signalError()
-		return false
-	}
-	if !c.useGas(tosca.Gas(InitCodeWordGas * tosca.SizeInWords(size.Uint64()))) {
-		c.status = statusOutOfGas
-		return false
+		return errOutOfGas
```

Comment:
This error is due to the size being greater than the MaxInitCodeSize, not due to out of gas.